lib/membrane_flac_plugin/membrane_flac_parser.ex

```diff
@@ -38,18 +47,43 @@ defmodule Membrane.FLAC.Parser do
     {[], state}
   end
 
+  defp calculate_pts(buffer, state) do
+    if state.generate_best_effort_timestamps? do
+      pts =
+        case buffer.metadata do
+          %{sample_rate: sample_rate, starting_sample_number: starting_sample_number} ->
+            (starting_sample_number / sample_rate * 1_000_000_000) |> trunc()
+
+          _credo_silencer ->
+            nil
+        end
+
+      Map.merge(buffer, %{pts: pts})
+    else
+      Map.merge(buffer, %{pts: state.input_pts})
+    end
+  end
+
   @impl true
-  def handle_buffer(:input, %Buffer{payload: payload}, _ctx, %{parser: parser} = state) do
+  def handle_buffer(
+        :input,
+        %Buffer{payload: payload, pts: input_pts},
+        _ctx,
+        %{parser: parser} = state
+      ) do
     case Engine.parse(payload, parser) do
       {:ok, results, parser} ->
         actions =
           results
           |> Enum.map(fn
-            %FLAC{} = format -> {:stream_format, {:output, format}}
-            %Buffer{} = buf -> {:buffer, {:output, buf}}
+            %FLAC{} = format ->
+              {:stream_format, {:output, format}}
+
+            %Buffer{} = buf ->
+              {:buffer, {:output, calculate_pts(buf, %{state | input_pts: input_pts})}}
           end)
 
-        {actions, %{state | parser: parser}}
+        {actions, %{state | parser: parser, input_pts: input_pts}}
```

Comment:
instead of setting `input_pts` two times, let's set it at the beginning of `handle_buffer`